### DISCUSSION TITLE
Purpose of the Emgu.CV nuget package

### DISCUSSION THREAD

**Discussion by sabuncu** [emgucv/emgucv#398] [Q&A] [answered]
On Windows 10 using Visual Studio 2019, I am able to develop and run the Hello World sample by installing *only* the **Emgu.CV.runtime.windows package (v4.4.0.4099).**  **I did not install the Emgu.CV** package and yet did not encounter any problems when developing/running my application.

I read the following explanation about the switch to the NetStandard 2.0 profile: [http://www.emgu.com/wiki/index.php/Version_History#Emgu.CV-4.2.0](http://www.emgu.com/wiki/index.php/Version_History#Emgu.CV-4.2.0).  But I really don't understand what that means for me - I am developing with .NET Framework 4.7.2.

Detailed explanations are welcome.  Thanks.

**Comment by emgucv** [comment, marked as answer]
Emgu.CV nuget package is a dependency for Emgu.CV.runtime.windows package, check the following page and expand the "Dependencies" tab:
https://www.nuget.org/packages/Emgu.CV.runtime.windows/

If you install Emgu.CV.runtime.windows nuget package (without explicitly installing Emgu.CV), the Emgu.CV nuget package will be automatically installed for you when nuget manager resolve the dependencies.

.NetStandard 2.0 profile is compatible with .NetFramework 4.6.1 and above. See support matrix below:
https://docs.microsoft.com/en-us/dotnet/standard/net-standard

If you are using .Net Framework 4.7.2, it is compatible with .NetStandard 2.0 profile used by the Emgu.CV nuget package.

**Comment by sabuncu** [comment]
Thank you for the clarification, you are right, I missed that dependency.

I would appreciate it if you could confirm my understanding:

The Emgu.CV.runtime.windows is dependent on Emgu.CV.

In turn, Emgu.CV is dependent on the following:

System.Drawing.Primitives
System.Runtime 
System.Runtime.InteropServices.RuntimeInformation

Also, Emgu.CV.runtime.windows lists the two following dependencies, but they are no longer made available:

Emgu.runtime.windows.msvc.rt.x64
Emgu.runtime.windows.msvc.rt.x86 

Thanks again.

**Reply by vpenades** [reply]
> Emgu.runtime.windows.msvc.rt.x64
> Emgu.runtime.windows.msvc.rt.x86

Yeah, I've just noticed the latest version of the Emgu.CV.runtime.windows package depends on these two, so why are they unlisted? it's just to "conceal" them?  if that's the case, What I've seen the common use case on dependency packages is to just add the case to the documentation of the package on nuget.org, as in "this package is not intended to be referenced directly, refer to package X"

**Comment by emgucv** [comment]
Unlisting a package hides it from search. This is an dependency package, it should not show up in nuget.org search results and should not be installed independently. 

Listing them in nuget.org 'pollutes' nuget search results. Think about searching for "Emgu" on nuget.org and serveral of those packages shows up, you spend way more time to search for the right package. We want only the important packages to show up in an "Emgu" search on nuget.org

**Reply by jleoutward** [reply]
Latest NuGet package v4.5.1.4349 install failed on VS2017 as well.  Error message:
"Severity	Code	Description	Project	File	Line	Suppression State
Error		Could not install package 'Emgu.runtime.windows.msvc.rt.x64 19.28.29336'. You are trying to install this package into a project that targets '.NETFramework,Version=v4.6.2', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author."

I tried to download and install "emgu.runtime.windows.msvc.rt.x64.19.28.29336.nupkg" directly but get the same error message above.

So bottom line is how can we upgrade to 4.5.1?
Would very appreciate any help you can provide.

**Reply by vpenades** [reply]
Emgu.CV is a Netstandard2.0 package. it needs to be installed against a compatible framework. This means that your project consuming the nuget package needs to be _at least_ net4.6.2, and recommended net4.7.2

**Reply by jleoutward** [reply]
My .net framework is 4.6.2 as indicated above.  I also just tried 4.8.  
Please note I'm trying to install Emgu.CV.runtime.windows.  Emgu.CV and Emgu.CV.Bitmap 4.5.1 install fine.

See this screenshot of VS2017:

![image](https://user-images.githubusercontent.com/78058619/106044717-f5313080-6094-11eb-885f-30c6d70b7ba9.png)



Here's the log in text form:
Attempting to gather dependency information for package 'Emgu.CV.runtime.windows.4.5.1.4349' with respect to project 'WindowsFormsApp1', targeting '.NETFramework,Version=v4.8'
Gathering dependency information took 351.52 ms
Attempting to resolve dependencies for package 'Emgu.CV.runtime.windows.4.5.1.4349' with DependencyBehavior 'Lowest'
Resolving dependency information took 0 ms
Resolving actions to install package 'Emgu.CV.runtime.windows.4.5.1.4349'
Resolved actions to install package 'Emgu.CV.runtime.windows.4.5.1.4349'
Retrieving package 'Emgu.CV.runtime.windows 4.5.1.4349' from 'nuget.org'.
  GET https://api.nuget.org/v3-flatcontainer/emgu.cv.runtime.windows/4.5.1.4349/emgu.cv.runtime.windows.4.5.1.4349.nupkg
Retrieving package 'Emgu.runtime.windows.msvc.rt.x64 19.28.29336' from 'nuget.org'.
Retrieving package 'Emgu.runtime.windows.msvc.rt.x86 19.28.29336' from 'nuget.org'.
  OK https://api.nuget.org/v3-flatcontainer/emgu.cv.runtime.windows/4.5.1.4349/emgu.cv.runtime.windows.4.5.1.4349.nupkg 344ms
Installing Emgu.CV.runtime.windows 4.5.1.4349.
Install failed. Rolling back...
Package 'Emgu.runtime.windows.msvc.rt.x64.19.28.29336' does not exist in project 'WindowsFormsApp1'
Package 'Emgu.runtime.windows.msvc.rt.x64.19.28.29336' does not exist in folder 'C:\Users\Outward\source\repos\WindowsFormsApp1\packages'
Executing nuget actions took 748.51 ms
Could not install package 'Emgu.runtime.windows.msvc.rt.x64 19.28.29336'. You are trying to install this package into a project that targets '.NETFramework,Version=v4.8', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.
Time Elapsed: 00:00:01.4386163

**Comment by sabuncu** [comment]
I am unable to follow this conversation as I am not familiar with the NuGet system, but really want to understand/learn.

> Unlisting a package hides it from search. 

If a package is unlisted, is it still installed?  For example, I don't see the two packages mentioned above installed on my system.

Sorry if these questions are too basic.

Update:

Came across these two links from Microsoft Docs, maybe it'll help others who are clueless like me:

[Deleting/unlisting packages](https://docs.microsoft.com/en-us/nuget/nuget-org/policies/deleting-packages)  
[Deprecating packages](https://docs.microsoft.com/en-us/nuget/nuget-org/deprecate-packages)

**Comment by emgucv** [comment]
@jleoutward 

You need to make sure that you have upgraded from "packages.config" to "Package reference".

Nuget dependency failed to resolve when using "packages.config" for nuget packages. Upgrading that to "Package reference" fixed that.

Detail can be found in this issue here: https://github.com/emgucv/emgucv/issues/399#issuecomment-767571013

**Comment by emgucv** [comment]
@sabuncu Yes, unlisted package can still be installed as a dependency. If nuget complains about "Emgu.runtime.windows.msvc.rt.x64", make sure your project use "Package Reference" instead of "packages.config".
